src/lib/logical_query_plan/lqp_utils.cpp

```diff
@@ -231,45 +231,75 @@ std::set<std::string> lqp_find_modified_tables(const std::shared_ptr<AbstractLQP
   return modified_tables;
 }
 
-std::shared_ptr<AbstractExpression> lqp_subplan_to_boolean_expression(
-    const std::shared_ptr<AbstractLQPNode>& begin, const std::optional<const std::shared_ptr<AbstractLQPNode>>& end) {
+namespace {
+/**
+ * Function creates a boolean expression from an lqp whilst preserving the same order of how expressions are evaluated.
+ *
+ *            LQP    --- lqp_subplan_to_boolean_expression(Predicate B) --->    boolean expression
+ *
+ *        Predicate B                                                    Predicate A       Predicate B
+ *             |                                                                \             /
+ *        Predicate A                                                             --- AND ---
+ *             |                                                                       |
+ *       Stored Table                                                          returned expression
+ *
+ * An lqp is evaluated from bottom to top. However, this function traverses an lqp from top to bottom. In the given lqp,
+ * Predicate B is reached first druing traversial, however, Predicate B can only be added to output expression once
+ * Predicate A was added as Predicate A is evaluated before Predicate B.
+ * The translated predicate expression from a visited predicate node is therefore not directly added to the output
+ * expression. Instead it is added once the next Predicate or Union node are translated. To allow this, the translated
+ * predicate expression is passed as the subsequent_expression parameter to the translation of its input nodes. If the
+ * subsequent_expression parameter is set, an `and` expression is created to combine the translated expression of the
+ * current node with the subsequent predicate expression.
+ */
+std::shared_ptr<AbstractExpression> _lqp_subplan_to_boolean_expression(
+    const std::shared_ptr<AbstractLQPNode>& begin, const std::optional<const std::shared_ptr<AbstractLQPNode>>& end,
+    const std::optional<const std::shared_ptr<AbstractExpression>>& subsequent_expression) {
   if (end && begin == *end) return nullptr;
 
-  static const auto whitelist =
-      std::set<LQPNodeType>{LQPNodeType::Projection, LQPNodeType::Sort, LQPNodeType::Validate, LQPNodeType::Limit};
-
-  if (whitelist.count(begin->type)) return lqp_subplan_to_boolean_expression(begin->left_input(), end);
```

Comment:
A lookup in the `whitelist` is not necessary as these node types have a case in the switch statement below.